File: Protocol/_theoreme_sim/src/generation/topologyGenerator.js
/* Purpose: Build connected geometric mesh topologies suitable for theorem simulation rounds. */

import { LAYOUT } from "../core/constants.js";
import { randomInRange } from "../core/random.js";
import {
  buildAdjacency,
  connectedComponents,
  distance,
  edgeKey,
} from "../core/graphUtils.js";

/**
 * @param {number} nodeCount
 * @param {number} linkRadius
 * @param {() => number} rng
 * @returns {{nodes:Array<{id:number,x:number,y:number}>, edges:Array<{a:number,b:number,distance:number,quality:number}>, adjacency:Map<number, Set<number>>}}
 */
export function generateConnectedTopology(nodeCount, linkRadius, rng) {
  const nodes = [];

  for (let id = 0; id < nodeCount; id += 1) {
    nodes.push({
      id,
      x: randomInRange(rng, LAYOUT.margin, LAYOUT.width - LAYOUT.margin),
      y: randomInRange(rng, LAYOUT.margin, LAYOUT.height - LAYOUT.margin),
    });
  }

  /** @type {Array<{a:number,b:number,distance:number,quality:number}>} */
  const edges = [];
  const existing = new Set();

  const addEdge = (a, b) => {
    if (a === b) {
      return;
    }
    const key = edgeKey(a, b);
    if (existing.has(key)) {
      return;
    }

    const d = distance(nodes[a], nodes[b]);
    const quality = Math.max(0.08, 1 - d / (linkRadius * 1.35));
    edges.push({ a: Math.min(a, b), b: Math.max(a, b), distance: d, quality });
    existing.add(key);
  };

  for (let i = 0; i < nodeCount; i += 1) {
    for (let j = i + 1; j < nodeCount; j += 1) {
      if (distance(nodes[i], nodes[j]) <= linkRadius) {
        addEdge(i, j);
      }
    }
  }

  // Ensure each node has at least one neighbor by linking it to its nearest node.
  let adjacency = buildAdjacency(nodeCount, edges);
  for (let i = 0; i < nodeCount; i += 1) {
    if (adjacency.get(i).size > 0) {
      continue;
    }

    let nearest = -1;
    let nearestDistance = Number.POSITIVE_INFINITY;
    for (let j = 0; j < nodeCount; j += 1) {
      if (i === j) {
        continue;
      }
      const d = distance(nodes[i], nodes[j]);
      if (d < nearestDistance) {
        nearestDistance = d;
        nearest = j;
      }
    }
    addEdge(i, nearest);
  }

  adjacency = buildAdjacency(nodeCount, edges);

  // Merge disconnected components by nearest cross-component edges.
  let components = connectedComponents(adjacency);
  while (components.length > 1) {
    const primary = components[0];
    const secondary = components[1];
    let bestA = primary[0];
    let bestB = secondary[0];
    let bestDistance = Number.POSITIVE_INFINITY;

    for (const a of primary) {
      for (const b of secondary) {
        const d = distance(nodes[a], nodes[b]);
        if (d < bestDistance) {
          bestDistance = d;
          bestA = a;
          bestB = b;
        }
      }
    }

    addEdge(bestA, bestB);
    adjacency = buildAdjacency(nodeCount, edges);
    components = connectedComponents(adjacency);
  }

  return { nodes, edges, adjacency };
}

File: Protocol/_theoreme_sim_py/src/generation/topology_generator.py
"""Purpose: Generate connected geometric topology used for theorem simulation rounds."""

from __future__ import annotations

from src.core.constants import LAYOUT
from src.core.graph_utils import (
    build_adjacency,
    connected_components,
    distance,
    edge_key,
)
from src.core.random import random_in_range


def generate_connected_topology(node_count: int, link_radius: float, rng) -> dict:
    nodes: list[dict] = []

    for node_id in range(node_count):
        nodes.append(
            {
                "id": node_id,
                "x": random_in_range(
                    rng,
                    float(LAYOUT["margin"]),
                    float(LAYOUT["width"] - LAYOUT["margin"]),
                ),
                "y": random_in_range(
                    rng,
                    float(LAYOUT["margin"]),
                    float(LAYOUT["height"] - LAYOUT["margin"]),
                ),
            }
        )

    edges: list[dict] = []
    existing: set[str] = set()

    def add_edge(a: int, b: int) -> None:
        if a == b:
            return

        key = edge_key(a, b)
        if key in existing:
            return

        d = distance(nodes[a], nodes[b])
        quality = max(0.08, 1 - d / (float(link_radius) * 1.35))
        edges.append(
            {
                "a": min(a, b),
                "b": max(a, b),
                "distance": d,
                "quality": quality,
            }
        )
        existing.add(key)

    for i in range(node_count):
        for j in range(i + 1, node_count):
            if distance(nodes[i], nodes[j]) <= link_radius:
                add_edge(i, j)

    adjacency = build_adjacency(node_count, edges)

    # Ensure every node has at least one neighbor.
    for i in range(node_count):
        if adjacency[i]:
            continue

        nearest = -1
        nearest_distance = float("inf")
        for j in range(node_count):
            if i == j:
                continue
            d = distance(nodes[i], nodes[j])
            if d < nearest_distance:
                nearest_distance = d
                nearest = j

        if nearest >= 0:
            add_edge(i, nearest)

    adjacency = build_adjacency(node_count, edges)

    # Merge components by shortest cross-edge.
    components = connected_components(adjacency)
    while len(components) > 1:
        primary = components[0]
        secondary = components[1]

        best_a = primary[0]
        best_b = secondary[0]
        best_distance = float("inf")

        for a in primary:
            for b in secondary:
                d = distance(nodes[a], nodes[b])
                if d < best_distance:
                    best_distance = d
                    best_a = a
                    best_b = b

        add_edge(best_a, best_b)
        adjacency = build_adjacency(node_count, edges)
        components = connected_components(adjacency)

    return {
        "nodes": nodes,
        "edges": edges,
        "adjacency": adjacency,
    }
